### ISSUE TITLE
Compiler panic with missing (inherited?) associated types

### ISSUE THREAD

**Issue by smangelsdorf** [rust-lang/rust#40789]
Two incomplete trait implementations triggered a compiler panic, when I was waiting for a compiler error to guide me on what to implement.

I've distilled my code down to this example:

```toml
[dependencies]
hyper = { git = "https://github.com/hyperium/hyper.git", ref = "43cf9aefe8f9147a0f08aa0f9d609d275549fd7c" }
```

```rust
// src/main.rs

extern crate hyper;

use hyper::client;

struct TestConnect {}

impl client::Service for TestConnect {}
impl client::Connect for TestConnect {}

fn main() {}
```

I expected to see a compile error, though I'm not sure what the error would / should have said.

Instead, this happened:

```
error: internal compiler error: /Users/rustbuild/src/rust-buildbot/slave/stable-dist-rustc-mac/build/src/librustc/traits/project.rs:1018: Tried to project an inherited associated type during coherence checking, which is currently not supported.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/stable-dist-rustc-mac/build/src/librustc_errors/lib.rs:375
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

## Meta

`rustc --version --verbose`:

```
rustc 1.16.0 (30cf806ef 2017-03-10)
binary: rustc
commit-hash: 30cf806ef8881c41821fbd43e5cf3699c5290c16
commit-date: 2017-03-10
host: x86_64-apple-darwin
release: 1.16.0
LLVM version: 3.9
```

Backtrace:

```
thread 'rustc' panicked at 'Box<Any>', /Users/rustbuild/src/rust-buildbot/slave/stable-dist-rustc-mac/build/src/librustc_errors/lib.rs:375
stack backtrace:
   1:        0x1100eba1c - std::sys::imp::backtrace::tracing::imp::write::h21ca2762819c7ae8
   2:        0x1100f837e - std::panicking::default_hook::{{closure}}::h38f99a37d00bb19b
   3:        0x1100f7f23 - std::panicking::default_hook::ha2186ee24b50729c
   4:        0x1100f8837 - std::panicking::rust_panic_with_hook::h979db19ee91d2a53
   5:        0x10cecd57b - std::panicking::begin_panic::h0a3d03ab7285de00
   6:        0x10d07a7e3 - rustc::session::opt_span_bug_fmt::{{closure}}::hcb2fbcb88915d137
   7:        0x10d07a2e9 - rustc::session::opt_span_bug_fmt::ha610cb4388904d76
   8:        0x10d079ec7 - rustc::session::span_bug_fmt::hb211c7ed93f0c716
   9:        0x10d0969b5 - rustc::traits::project::assemble_candidates_from_impls::{{closure}}::hc7fec52b4de1e221
  10:        0x10d09639b - rustc::traits::project::assemble_candidates_from_impls::h0c88b5ddc1bbff32
  11:        0x10d0944d2 - rustc::traits::project::opt_normalize_projection_type::h3e64f886bc77e0ad
  12:        0x10d08fa16 - rustc::traits::project::project_and_unify_type::h45efdef6f0383718
  13:        0x10cfe5355 - rustc::infer::InferCtxt::commit_if_ok::h486cc5a9768c1f24
  14:        0x10d09c013 - rustc::traits::select::SelectionContext::evaluate_predicate_recursively::h1d78e7fcef08e60d
  15:        0x10d09b25d - rustc::traits::select::SelectionContext::probe::ha1332923e5778ff7
  16:        0x10d07adaa - rustc::traits::coherence::overlapping_impls::h6de55c0b84018eb4
  17:        0x10d0aba22 - rustc::traits::specialize::specialization_graph::Graph::insert::hc3cc56437226e6fe
  18:        0x10d0e62c2 - rustc::ty::trait_def::TraitDef::add_impl_for_specialization::h5814609b0e1e457b
  19:        0x10c3e5c77 - <rustc_typeck::coherence::overlap::OverlapChecker<'cx, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'v>>::visit_item::h3898b082e3a75942
  20:        0x10c3ead1e - rustc_typeck::coherence::check_coherence::h53730d6e44c7dc36
  21:        0x10c3f45b0 - rustc_typeck::check_crate::h4e57307efed0103d
  22:        0x10bfd0db0 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h9e118a56f5957bd6
  23:        0x10bfb3367 - rustc_driver::driver::phase_3_run_analysis_passes::h0569f99df4eb53d4
  24:        0x10bfa2a90 - rustc_driver::driver::compile_input::hb50de2aa3bf90e59
  25:        0x10bfea48e - rustc_driver::run_compiler::h4b71cb0b28ba1b4a
  26:        0x10bf01ff8 - std::panicking::try::do_call::h7cd9045c0206869e
  27:        0x1100fb44a - __rust_maybe_catch_panic
  28:        0x10bf2a913 - <F as alloc::boxed::FnBox<A>>::call_box::hca1d0347a9530a7f
  29:        0x1100f74a4 - std::sys::imp::thread::Thread::new::thread_start::hca57ad22492f366f
  30:     0x7fffb9f6daaa - _pthread_body
  31:     0x7fffb9f6d9f6 - _pthread_start
```

**Comment by smangelsdorf** [comment]
This is fixed in Rust 1.19.0, and I now get the error I was expecting:

```
error[E0119]: conflicting implementations of trait `hyper::client::Connect` for type `TestConnect`:
 --> src/main.rs:8:1
  |
8 | impl client::Connect for TestConnect {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `hyper`
```
